server/sonar-server/src/main/java/org/sonar/server/qualitygate/QgateProjectFinder.java

```diff
@@ -19,22 +19,92 @@
  */
 package org.sonar.server.qualitygate;
 
+import com.google.common.base.Function;
+import com.google.common.base.Predicate;
+import com.google.common.collect.Lists;
+import java.util.Collection;
+import java.util.List;
+import javax.annotation.Nonnull;
 import org.sonar.api.server.ServerSide;
+import org.sonar.api.utils.Paging;
+import org.sonar.api.web.UserRole;
+import org.sonar.db.DbClient;
+import org.sonar.db.DbSession;
 import org.sonar.db.qualitygate.ProjectQgateAssociation;
 import org.sonar.db.qualitygate.ProjectQgateAssociationDao;
 import org.sonar.db.qualitygate.ProjectQgateAssociationDto;
 import org.sonar.db.qualitygate.ProjectQgateAssociationQuery;
 import org.sonar.db.qualitygate.QualityGateDao;
 import org.sonar.db.qualitygate.QualityGateDto;
 import org.sonar.server.exceptions.NotFoundException;
+import org.sonar.server.user.UserSession;
 
-import java.util.List;
-
+import static com.google.common.collect.FluentIterable.from;
 import static com.google.common.collect.Lists.newArrayList;
+import static org.sonar.api.utils.Paging.forPageIndex;
 
 @ServerSide
 public class QgateProjectFinder {
 
+  private final DbClient dbClient;
+  private final QualityGateDao qualitygateDao;
+  private final ProjectQgateAssociationDao associationDao;
+  private final UserSession userSession;
+
+  public QgateProjectFinder(DbClient dbClient, UserSession userSession) {
+    this.dbClient = dbClient;
+    this.userSession = userSession;
+    this.qualitygateDao = dbClient.qualityGateDao();
+    this.associationDao = dbClient.projectQgateAssociationDao();
+  }
+
+  public Association find(ProjectQgateAssociationQuery query) {
+    DbSession dbSession = dbClient.openSession(false);
+    try {
+      validateId(dbSession, query.gateId());
+      List<ProjectQgateAssociationDto> projects = associationDao.selectProjects(dbSession, query);
+      List<ProjectQgateAssociationDto> authorizedProjects = keepAuthorizedProjects(dbSession, projects);
+
+      Paging paging = forPageIndex(query.pageIndex()).withPageSize(query.pageSize()).andTotal(authorizedProjects.size());
+      return new Association(toProjectAssociations(getPaginatedProjects(authorizedProjects, paging)), paging.hasNextPage());
+    } finally {
+      dbClient.closeSession(dbSession);
+    }
+  }
+
+  private Long validateId(DbSession dbSession, String gateId) {
+    QualityGateDto qualityGateDto = qualitygateDao.selectById(dbSession, Long.valueOf(gateId));
+    if (qualityGateDto == null) {
+      throw new NotFoundException("Quality gate '" + gateId + "' does not exists.");
+    }
+    return qualityGateDto.getId();
+  }
+
+  private static List<ProjectQgateAssociationDto> getPaginatedProjects(List<ProjectQgateAssociationDto> projects, Paging paging) {
```

Comment:
even better, use something similar to: https://github.com/SonarSource/sonarqube/blob/0d04b2f74fd9ab532b3dc82f4ebeebac8b2cca4f/server/sonar-server/src/main/java/org/sonar/server/measure/ws/ComponentTreeDataLoader.java#L283-L283